packages/assets-controllers/src/RatesController/RatesController.ts

```diff
@@ -0,0 +1,197 @@
+   * @param options.state - Initial state to set on this controller.
+   * @param options.fetchMultiExchangeRate - Fetches the exchange rate from an external API. This option is primarily meant for use in unit tests.
+   * @param options.onStart - Optional callback to be executed when the polling stops.
+   * @param options.onStop - Optional callback to be executed when the polling starts.
+   */
+  constructor({
+    interval = 180000,
+    messenger,
+    state,
+    includeUsdRate,
+    fetchMultiExchangeRate = defaultFetchExchangeRate,
+    onStart,
+    onStop,
+  }: RatesControllerArgs) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.#includeUsdRate = includeUsdRate;
+    this.#fetchMultiExchangeRate = fetchMultiExchangeRate;
+    this.#onStart = onStart;
+    this.#onStop = onStop;
+    this.#setIntervalLength(interval);
+  }
+
+  /**
+   * Sets the interval length for polling.
+   *
+   * @param intervalLength - The length of the interval in milliseconds.
+   */
+  #setIntervalLength(intervalLength: number) {
+    this.#intervalLength = intervalLength;
+  }
+
+  /**
+   * Updates the BTC rates by fetching new data.
+   */
+  async updateRates(): Promise<void> {
+    const releaseLock = await this.#mutex.acquire();
+    const { currency, cryptocurrencyList } = this.state;
+
+    try {
+      const response = await this.#fetchMultiExchangeRate(
+        currency,
+        cryptocurrencyList,
+        this.#includeUsdRate,
+      );
+
+      const updatedRates: ConversionRates = {};
+      for (const [key, value] of Object.entries(response)) {
+        updatedRates[key] = {
+          conversionDate: Date.now() / 1000,
+          conversionRate: value[currency],
+          usdConversionRate: value.usd || null,
+        };
+      }
+
+      this.update(() => {
+        return {
+          ...this.state,
+          rates: updatedRates,
+        };
+      });
+    } finally {
+      releaseLock();
+    }
+  }
+
+  /**
+   * Executes the polling operation to update rates.
+   */
+  async #executePoll(): Promise<void> {
+    await this.updateRates();
+  }
+
+  /**
+   * Starts the polling process.
+   */
+  async start(): Promise<void> {
+    if (this.#intervalId) {
+      return;
+    }
+
+    await this.#onStart?.();
+    this.#intervalId = setInterval(() => {
+      this.#executePoll().catch(console.error);
+    }, this.#intervalLength);
+  }
+
+  /**
+   * Stops the polling process.
+   */
+  async stop(): Promise<void> {
+    if (!this.#intervalId) {
+      return;
+    }
+
+    clearInterval(this.#intervalId);
+    this.#intervalId = undefined;
+    await this.#onStop?.();
+  }
+
+  getCryptocurrencyList(): string[] {
+    const { cryptocurrencyList } = this.state;
+    return cryptocurrencyList;
+  }
+
+  setCryptocurrencyList(list: string[]): void {
+    this.update(() => {
+      return {
+        ...this.state,
+        cryptocurrencyList: list,
```

Comment:
I wanted to provide a simple way to update this list in case we need to fetch rates for other currencies, as we want to support other non-EVM blockchains, I'm thinking how can we get the data for SOL or STRK.

I'm also unsure if we have a pattern for this already established, let me know if you have any suggestions.

For the case of updating the state while executing an iteration, I missed setting a mutex for this method which is now done in [aa22c5d](https://github.com/MetaMask/core/pull/4242/commits/aa22c5d7cec9aa9938ab31d3cc3a419be50b6a8c).